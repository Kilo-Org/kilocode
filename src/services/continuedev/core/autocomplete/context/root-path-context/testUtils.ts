import fs from "fs"
import path from "path"
import { fileURLToPath } from "node:url"
import { expect, vi } from "vitest"

const __dirname = path.dirname(fileURLToPath(import.meta.url))

import Parser from "web-tree-sitter"
import type { Position } from "../../../index.js"
import { testIde } from "../../../test/fixtures"
import { getAst, getTreePathAtCursor } from "../../util/ast"
import { ImportDefinitionsService } from "../ImportDefinitionsService"
import { RootPathContextService } from "./RootPathContextService"

function splitTextAtPosition(fileContent: string, position: Position): [string, string] {
	const lines = fileContent.split("\n")
	let currentPos = 0

	// Calculate position based on the provided line and character
	for (let i = 0; i < position.line; i++) {
		currentPos += lines[i].length + 1 // +1 for the newline character
	}
	const splitPos = currentPos + position.character

	return [fileContent.slice(0, splitPos), fileContent.slice(splitPos)]
}

export async function testRootPathContext(
	folderName: string,
	relativeFilepath: string,
	position: Position,
	expectedDefinitionPositions: Parser.Point[],
) {
	// Create a mocked instance of RootPathContextService
	const ide = testIde
	const importDefinitionsService = new ImportDefinitionsService(ide)
	const service = new RootPathContextService(importDefinitionsService, ide)

	const getSnippetsMock = vi
		// @ts-expect-error -- spying on private method for test verification
		.spyOn(service, "getSnippets")
		// @ts-expect-error -- mocking private method implementation
		.mockImplementation((_filePath, _endPosition) => {
			return []
		})

	// Copy the folder to the test directory
	const folderPath = path.join(__dirname, "__fixtures__", folderName)
	let workspaceDir = (await ide.getWorkspaceDirs())[0] as string
	if (workspaceDir.startsWith("file:")) {
		workspaceDir = fileURLToPath(workspaceDir)
	}
	const testFolderPath = path.join(workspaceDir, folderName)
	fs.cpSync(folderPath, testFolderPath, {
		recursive: true,
		force: true,
	})

	// Get results of root path context
	const startPath = relativeFilepath.startsWith("file:")
		? fileURLToPath(relativeFilepath)
		: path.join(testFolderPath, relativeFilepath)
	const [prefix, suffix] = splitTextAtPosition(fs.readFileSync(startPath, "utf8"), position)
	const fileContents = prefix + suffix
	const ast = await getAst(startPath, fileContents)
	if (!ast) {
		throw new Error("AST is undefined")
	}

	const treePath = await getTreePathAtCursor(ast, prefix.length)
	await service.getContextForPath(startPath, treePath)

	expect(getSnippetsMock).toHaveBeenCalledTimes(expectedDefinitionPositions.length)

	expectedDefinitionPositions.forEach((position, index) => {
		expect(getSnippetsMock).toHaveBeenNthCalledWith(
			index + 1,
			expect.any(String), // filepath argument
			position,
			expect.any(String), // language argument
		)
	})
}
