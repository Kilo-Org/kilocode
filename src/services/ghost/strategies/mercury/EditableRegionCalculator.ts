import * as vscode from "vscode"
import { countTokens } from "../../utils/tokenHelpers"

/**
 * Result of calculating an editable region
 */
export interface EditableRegionResult {
	content: string // The actual text content
	startLine: number // Starting line in document (0-based)
	endLine: number // Ending line in document (0-based)
	tokensUsed: number // Actual tokens used
}

/**
 * Legacy interface for backward compatibility with tests
 * @deprecated Use EditableRegionResult instead
 */
export interface EditableRegion {
	editableStart: number
	editableEnd: number
	totalLines: number
	tokensUsed: number
}

/**
 * Simplified editable region calculator that returns content and positions in one call.
 * Uses proper tiktoken-based token counting.
 */
export class EditableRegionCalculator {
	/**
	 * Calculate editable region and return content + positions in one call.
	 * This is what Mercury strategy actually needs - no separate extract step.
	 */
	public calculateEditableRegionContent(
		document: vscode.TextDocument,
		cursorRange: vscode.Range,
		maxTokens: number = 512,
	): EditableRegionResult {
		const fileLines = document.getText().split("\n")
		const cursorLine = cursorRange.start.line

		console.log("EditableRegionCalculator DEBUG:")
		console.log("  fileLines.length:", fileLines.length)
		console.log("  cursorLine:", cursorLine)
		console.log(
			"  fileLines:",
			fileLines.map((line, i) => `${i}: "${line}"`),
		)

		// Start with just the cursor line
		let startLine = cursorLine
		let endLine = cursorLine
		let currentContent = fileLines[cursorLine] || ""
		let totalTokens = countTokens(currentContent)

		console.log("  initial currentContent:", JSON.stringify(currentContent))

		// Expand alternately above and below cursor
		let expandAbove = true
		while (totalTokens < maxTokens) {
			let addedLine = false

			if (expandAbove) {
				if (startLine > 0) {
					const candidateLine = fileLines[startLine - 1]
					const candidateTokens = countTokens(candidateLine)

					if (totalTokens + candidateTokens <= maxTokens) {
						startLine--
						currentContent = candidateLine + "\n" + currentContent
						totalTokens += candidateTokens
						addedLine = true
						console.log(`  expanded above to line ${startLine}: "${candidateLine}"`)
					}
				}
			} else {
				if (endLine < fileLines.length - 1) {
					const candidateLine = fileLines[endLine + 1]
					const candidateTokens = countTokens(candidateLine)

					if (totalTokens + candidateTokens <= maxTokens) {
						endLine++
						currentContent = currentContent + "\n" + candidateLine
						totalTokens += candidateTokens
						addedLine = true
						console.log(`  expanded below to line ${endLine}: "${candidateLine}"`)
					}
				}
			}

			if (!addedLine) break
			// CRITICAL FIX: Always try to include the final closing brace if we're missing it
			// This ensures we capture complete code blocks
			if (endLine < fileLines.length - 1) {
				const nextLine = fileLines[endLine + 1]
				if (nextLine.trim() === "}" || nextLine.trim() === "};") {
					const extraTokens = countTokens(nextLine)
					// Include closing brace even if it slightly exceeds token limit
					if (totalTokens + extraTokens <= maxTokens + 10) {
						// Small buffer for closing braces
						endLine++
						currentContent = currentContent + "\n" + nextLine
						totalTokens += extraTokens
						console.log(`  included closing brace at line ${endLine}: "${nextLine}"`)
					}
				}
			}
			expandAbove = !expandAbove
		}

		console.log("  final startLine:", startLine, "endLine:", endLine)
		console.log("  final currentContent:", JSON.stringify(currentContent))

		return {
			content: currentContent,
			startLine,
			endLine,
			tokensUsed: totalTokens,
		}
	}

	/**
	 * Legacy method for backward compatibility with tests
	 * @deprecated Use calculateEditableRegionContent instead
	 */
	public calculateEditableRegion(
		document: vscode.TextDocument,
		cursorRange: vscode.Range,
		maxTokens: number = 512,
	): EditableRegion {
		const result = this.calculateEditableRegionContent(document, cursorRange, maxTokens)
		return {
			editableStart: result.startLine,
			editableEnd: result.endLine,
			totalLines: document.lineCount,
			tokensUsed: result.tokensUsed,
		}
	}

	/**
	 * Legacy method for backward compatibility with tests
	 * @deprecated Use calculateEditableRegionContent instead
	 */
	public extractEditableContent(document: vscode.TextDocument, region: EditableRegion): string {
		const lines = document.getText().split("\n")
		return lines.slice(region.editableStart, region.editableEnd + 1).join("\n")
	}
}
