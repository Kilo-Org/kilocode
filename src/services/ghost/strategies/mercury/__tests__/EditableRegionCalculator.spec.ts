import { describe, it, expect } from "vitest"
import { EditableRegionCalculator } from "../EditableRegionCalculator"
import { parseContentWithCursor } from "./testUtils"

describe("EditableRegionCalculator", () => {
	let calculator: EditableRegionCalculator

	beforeEach(() => {
		calculator = new EditableRegionCalculator()
	})

	describe("calculateEditableRegion", () => {
		it("should expand around cursor with high token limit", () => {
			const contentWithCursor = `line 0
line 1
line 2 - cu␣rsor here
line 3
line 4`

			const { document, cursorRange } = parseContentWithCursor(contentWithCursor)
			const region = calculator.calculateEditableRegion(document, cursorRange, 100)

			expect(region).toMatchSnapshot()
		})

		it("should respect low token limits", () => {
			const contentWithCursor = `line 0
line 1
line 2 - cu␣rsor here
line 3
line 4`

			const { document, cursorRange } = parseContentWithCursor(contentWithCursor)
			const region = calculator.calculateEditableRegion(document, cursorRange, 10)

			expect(region).toMatchSnapshot()
		})

		it("should handle single line file", () => {
			const contentWithCursor = `only li␣ne`

			const { document, cursorRange } = parseContentWithCursor(contentWithCursor)
			const region = calculator.calculateEditableRegion(document, cursorRange, 100)

			expect(region).toMatchSnapshot()
		})
	})

	describe("extractEditableContent", () => {
		it("should extract multi-line region content", () => {
			const { document } = parseContentWithCursor(`line 0
line ␣1
line 2
line 3
line 4`)

			const region = { editableStart: 1, editableEnd: 3, totalLines: 5, tokensUsed: 50 }
			const extracted = calculator.extractEditableContent(document, region)

			expect(extracted).toMatchSnapshot()
		})

		it("should extract single line region content", () => {
			const { document } = parseContentWithCursor(`line 0
target line with ␣special chars & symbols
line 2`)

			const region = { editableStart: 1, editableEnd: 1, totalLines: 3, tokensUsed: 20 }
			const extracted = calculator.extractEditableContent(document, region)

			expect(extracted).toMatchSnapshot()
		})
	})
})
