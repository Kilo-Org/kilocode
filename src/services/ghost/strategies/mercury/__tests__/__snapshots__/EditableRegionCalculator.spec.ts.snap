// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableRegionCalculator > calculateEditableRegion > should expand around cursor with high token limit 1`] = `
{
  "editableEnd": 4,
  "editableStart": 0,
  "tokensUsed": 13,
  "totalLines": 5,
}
`;

exports[`EditableRegionCalculator > calculateEditableRegion > should handle single line file 1`] = `
{
  "editableEnd": 0,
  "editableStart": 0,
  "tokensUsed": 3,
  "totalLines": 1,
}
`;

exports[`EditableRegionCalculator > calculateEditableRegion > should respect low token limits 1`] = `
{
  "editableEnd": 3,
  "editableStart": 0,
  "tokensUsed": 11,
  "totalLines": 5,
}
`;

exports[`EditableRegionCalculator > extractEditableContent > should extract multi-line region content 1`] = `
"line 1
line 2
line 3"
`;

exports[`EditableRegionCalculator > extractEditableContent > should extract single line region content 1`] = `"target line with special chars & symbols"`;
